src/ffi/web_socket/error.rs

```diff
@@ -0,0 +1,272 @@
+    #[error("UTF-8 encoding error")]
+    Utf8,
+    /// Invalid URL.
+    #[error("URL error: {url_error}")]
+    Url {
+        /// The URL error as a string
+        url_error: String,
+    },
+    /// HTTP error.
+    #[error("HTTP error: {}", response.status_code)]
+    Http {
+        /// Error response from the server.
+        response: http::Response,
+    },
+    /// HTTP format error.
+    #[error("HTTP format error: {error}")]
+    HttpFormat {
+        /// HTTP format error.
+        error: http::HttpError,
+    },
+}
+impl From<&tokio_tungstenite::tungstenite::Error> for WebSocketError {
+    fn from(tungstenite_error: &tokio_tungstenite::tungstenite::Error) -> Self {
+        match tungstenite_error {
+            tokio_tungstenite::tungstenite::Error::ConnectionClosed => Self::ConnectionClosed,
+            tokio_tungstenite::tungstenite::Error::AlreadyClosed => Self::AlreadyClosed,
+            tokio_tungstenite::tungstenite::Error::Io(io_error) => Self::Io {
+                io_error: io_error.into(),
+            },
+            tokio_tungstenite::tungstenite::Error::Tls(tls_error) => Self::Tls {
+                tls_error: tls_error.to_string(),
+            },
+            tokio_tungstenite::tungstenite::Error::Capacity(capacity_error) => Self::Capacity {
+                capacity_error: capacity_error.into(),
+            },
+            tokio_tungstenite::tungstenite::Error::Protocol(protocol_error) => Self::Protocol {
+                protocol_error: protocol_error.into(),
+            },
+            tokio_tungstenite::tungstenite::Error::SendQueueFull(message) => Self::SendQueueFull {
+                message: message.into(),
+            },
+            tokio_tungstenite::tungstenite::Error::Utf8 => Self::Utf8,
+            tokio_tungstenite::tungstenite::Error::Url(url_error) => Self::Url {
+                url_error: url_error.to_string(),
+            },
+            tokio_tungstenite::tungstenite::Error::Http(response) => Self::Http {
+                response: response.into(),
+            },
+            tokio_tungstenite::tungstenite::Error::HttpFormat(error) => Self::HttpFormat {
+                error: error.into(),
+            },
+        }
+    }
+}
+
+/// [tungstenite::error::CapacityError], but with `uniffi` support.
+/// Indicates the specific type/cause of a capacity error.
+#[derive(Debug, PartialEq, Eq, Clone, Copy, thiserror::Error, uniffi::Error)]
+pub enum CapacityError {
+    /// Too many headers provided (see [`httparse::Error::TooManyHeaders`]).
+    #[error("Too many headers")]
+    TooManyHeaders,
+    /// Received header is too long.
+    /// Message is bigger than the maximum allowed size.
+    #[error("Message too long: {size} > {max_size}")]
+    MessageTooLong {
+        /// The size of the message.
+        size: u64,
+        /// The maximum allowed message size.
+        max_size: u64,
+    },
+}
+impl From<&tokio_tungstenite::tungstenite::error::CapacityError> for CapacityError {
```

Comment:
Since I suggested it on the other enums, I think `use tokio_tungstenite::tungstenite::error::CapacityError as TungsteniteCapacityError` and then use the aliased import is cleaner.